### PR TITLE
[wasm-ep] Use DOTNET_DiagnosticPorts to configure Diagnostic Server

### LINKED ISSUE #73011: [wasm-ep] Configure the Diagnostic Server using DOTNET_DiagnosticPorts
In order to make diagnostics work with Blazor, we need to provide a mechanism that works with their WebAssebly startup sequence (near here: [MonoPlatform.ts](https://github.com/dotnet/aspnetcore/blob/af7830f81669011ee5e6625d7fbe9f3384aa55a3/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts#L304)).  Using a `MonoConfig` `diagnostic_options` section doesn't work because they don't have this object.  Additionally the current call to `mono_wasm_init_diagnostics` is in `mono_wasm_before_user_runtime_initialized` which is not called for Blazor.

The other constraint is that `mono_wasm_init_diagnostics` is async (because it waits for the DS thread to start), so we need to call it during an async part of runtime startup.

---

So the task is:

1. Use the `DOTNET_DiagnosticPorts` environment variable to configure the diagnostic server - similar to desktop.  Setting it to a value like `ws://127.0.0.1:8088/diagnostics,connect,suspend` will cause the runtime to connect to the given WebSocket URL and suspend until a diagnostic tool sends a resume command.  We will only support `connect`, not `listen` (which is not meaningful for a websocket). We will support `suspend` (the default) and `nosuspend`
2. Blazor can setenv `DOTENT_DiagnosticPorts` from [their `onRuntimeInitialized` callback](https://github.com/dotnet/aspnetcore/blob/af7830f81669011ee5e6625d7fbe9f3384aa55a3/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts#L326)
3. Move the call to `await mono_wasm_init_diagnostics()` to `mono_wasm_after_user_runtime_initialized` which is async and runs after `onRuntimeInitialized` but before we call [Blazor's `postRun`](https://github.com/dotnet/aspnetcore/blob/af7830f81669011ee5e6625d7fbe9f3384aa55a3/src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts#L451) (which calls `MONO.mono_wasm_load_runtime` which is a sync function that actually begins running our C code - ie we must have the diagnostic server running by this point)

### PR DESCRIPTION
**Pull request by lambdageek**
Parse it the same way that the C code does:

```
   <uri>[,<connect|listen>][,<suspend|nosuspend>]
```

- uri should be a websocket uri
- listen is not supported as it doesn't make sense with a WebSocket
- connect is the default if omitted
- suspend is the default if omitted

---

Additionally, move `mono_wasm_diagnostics_init` to later in the
startup flow.  This gives Blazor a chance to set
`DOTNET_DiagnosticPorts` from their `onRuntimeInitialized` callback.
An unfortunate problem here is that blazor and non-blazor startup need to call diagnostics init from different places because of how environment variables are populated.  So we try to init diagnostics twice and set a flag.

Fixes https://github.com/dotnet/runtime/issues/73011

### CODE REVIEW

File: src/mono/wasm/runtime/startup.ts [hunk 1]
```diff
@@ -329,6 +331,10 @@ async function mono_wasm_after_user_runtime_initialized(): Promise<void> {
 
         if (runtimeHelpers.diagnosticTracing) console.debug("MONO_WASM: Initializing mono runtime");
 
+        if (MonoWasmThreads) {
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/startup.ts:334]: maybe you could now move it earlier in the startup, just after `beforeOnRuntimeInitialized.promise_control.resolve`. but I'm not sure it's useful.

**Comment by pavelsavara** [comment]
Please also update `src\mono\sample\wasm\browser-eventpipe\main.js` to use the env variable
Ideally also replacing `BINDING.bind_static_method` with `[JSExport]`
or is @maraf already looking at that ?

**Comment by maraf** [comment]
> Ideally also replacing `BINDING.bind_static_method` with `[JSExport]`
> or is @maraf Marek Fisera FTE already looking at that ?

Rewrite to JSExport is here https://github.com/dotnet/runtime/pull/73367

**Comment by lambdageek** [comment]
@pavelsavara @maraf  I'd like to get https://github.com/dotnet/runtime/pull/73305 in before this one.  I can't really test what's here until that other PR goes on.

@maraf I think I want to delete of `browser-mt-eventpipe`. It was just a proof of concept.  `browser-eventpipe` is good (and i'll update it to use `DOTNET_DiagnosticPorts`) and the new `browser-threads` in #73305 is good.  Both of them should be able to run in CI, I think.  `browser-mt-eventpipe` isn't really needed.

File: src/mono/wasm/runtime/startup.ts (hunk 1 above)
Comment (lambdageek) [reply src/mono/wasm/runtime/startup.ts:334]: I'll move it. the only hard requirement is that the diagnostic server has to start before we call into driver.c to start the root domain initialization.  Starting later is better

Comment (lambdageek) [reply src/mono/wasm/runtime/startup.ts:334]: Ah, there's an annoying problem here.  In the non-Blazor startup we don't set environment variables until `_apply_configuration_from_args`.  So actually I have to try to init diagnostics in two places.

File: src/mono/wasm/runtime/memory.ts [hunk 2]
```diff
@@ -263,6 +263,18 @@ export function mono_wasm_load_bytes_into_heap(bytes: Uint8Array): VoidPtr {
     return memoryOffset;
 }
 
+export function getEnv(name: string): string | null {
+    let charPtr: CharPtr = <any>0;
+    try {
+        charPtr = cwraps.mono_wasm_getenv(name);
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/memory.ts:269]: Is there any memory to free? (what owns the buffer?)

File: src/mono/wasm/runtime/diagnostics/index.ts [hunk 3]
```diff
@@ -94,12 +94,25 @@ export const diagnostics: Diagnostics = getDiagnostics();
 let suspendOnStartup = false;
 let diagnosticsServerEnabled = false;
 
-export async function mono_wasm_init_diagnostics(options: DiagnosticOptions): Promise<void> {
+let diagnosticsInitialized = false;
+
+export async function mono_wasm_init_diagnostics(opts: "env" | DiagnosticOptions): Promise<void> {
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/diagnostics/index.ts:99]: is the signature with `DiagnosticOptions` param type still useful ?

File: src/mono/wasm/runtime/memory.ts (hunk 2 above)
Comment (pavelsavara) [reply src/mono/wasm/runtime/memory.ts:269]: my bad

**Review by pavelsavara — APPROVED** [review]

File: src/mono/wasm/runtime/diagnostics/index.ts (hunk 3 above)
Comment (lambdageek) [reply src/mono/wasm/runtime/diagnostics/index.ts:99]: No. I want to do a follow up PR to simplify more of the diagnostics support. removing this argument is just one piece.

**Review by maraf — APPROVED** [review]

**Pull request merged**